clang-tools-extra/clang-tidy/readability/UseBuiltinLiteralsCheck.cpp

```diff
@@ -0,0 +1,171 @@
+                                          integerLiteral().bind("int"),
+                                          floatLiteral().bind("float"))))))
+                   .bind("expr")),
+      this);
+  Finder->addMatcher(
+      traverse(TK_IgnoreUnlessSpelledInSource,
+               explicitCastExpr(has(initListExpr(has(ignoringParenImpCasts(
+                                    anyOf(characterLiteral().bind("char"),
+                                          integerLiteral().bind("int"),
+                                          floatLiteral().bind("float")))))))
+                   .bind("expr")),
+      this);
+}
+
+void UseBuiltinLiteralsCheck::check(const MatchFinder::MatchResult &Result) {
+
+  const auto &Sources = *Result.SourceManager;
+  const auto &Nodes = Result.Nodes;
+
+  const auto *MatchedCast = Nodes.getNodeAs<ExplicitCastExpr>("expr");
+  assert(MatchedCast);
+
+  std::string CastType = MatchedCast->getTypeAsWritten().getAsString();
+  std::string Fix; // Replacement string for the fix-it hint.
+  StringRef Seq;   // Literal sequence, prefix or suffix.
+
+  if (const auto *CharLit = Nodes.getNodeAs<CharacterLiteral>("char");
+      CharLit && CharPrefix.contains(CastType)) {
+    if (const auto &Rep = CharPrefix.at(CastType); Rep(getLangOpts())) {
+
+      StringRef LitText =
+          getRawStringRef(CharLit->getLocation(), Sources, getLangOpts());
+
+      Seq = Rep.Seq;
+      if (!CharLit->getLocation().isMacroID()) {
+        Fix.append(Rep.Seq);
+        Fix.append(CharRegex.sub("", LitText.str()));
+      }
+    }
+  } else if (const auto *IntLit = Nodes.getNodeAs<IntegerLiteral>("int");
+             IntLit && IntSuffix.contains(CastType)) {
+    if (const auto &Rep = IntSuffix.at(CastType); Rep(getLangOpts())) {
+
+      StringRef LitText =
+          getRawStringRef(IntLit->getLocation(), Sources, getLangOpts());
+
+      Seq = Rep.Seq;
+      if (!IntLit->getLocation().isMacroID()) {
+        Fix.append(IntRegex.sub("", LitText.str()));
+        Fix.append(Rep.Seq);
+      }
+    }
+  } else if (const auto *FloatLit = Nodes.getNodeAs<FloatingLiteral>("float");
+             FloatLit && FloatSuffix.contains(CastType)) {
+    if (const auto &Rep = FloatSuffix.at(CastType); Rep(getLangOpts())) {
+
+      StringRef LitText =
+          getRawStringRef(FloatLit->getLocation(), Sources, getLangOpts());
+
+      Seq = Rep.Seq;
+      if (!FloatLit->getLocation().isMacroID()) {
+        Fix.append(FloatRegex.sub("", LitText.str()));
+        Fix.append(Rep.Seq);
+      }
+    }
+  }
+
+  const TypeLoc CastTypeLoc = MatchedCast->getTypeInfoAsWritten()->getTypeLoc();
+
+  if (!Fix.empty() && !CastTypeLoc.getBeginLoc().isMacroID()) {
+
+    // Recommend fix-it when no part of the explicit cast comes from a macro.
+    diag(MatchedCast->getExprLoc(), "use builtin literals instead of casts")
+        << FixItHint::CreateReplacement(MatchedCast->getSourceRange(),
+                                        Fix.c_str());
```

Comment:
`Fix` is a local variable. Isn't this `c_str()` escaping some to-be-destructed memory area here? It seems `FixItHint`'s factories can be created from a `std::string` normally.

Reply:
I overlooked the fact that the `StringRef` argument may also be constructed from a `const std::string&`. Looking at the lifetime of `Fix`, it doesn't seem like the pointer would be invalidated before it's used, it's just not as terse.